src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs

```diff
@@ -34,47 +35,92 @@ public static ImmutableQueue<T> Create<T>()
         [Pure]
         public static ImmutableQueue<T> Create<T>(T item)
         {
-            return ImmutableQueue<T>.Empty.Enqueue(item);
+            return new ImmutableQueue<T>(ImmutableStack.Create(item), ImmutableStack<T>.Empty);
         }
 
         /// <summary>
-        /// Creates a new immutable collection prefilled with the specified items.
+        /// Creates a new immutable queue from the specified items.
         /// </summary>
-        /// <typeparam name="T">The type of items stored by the collection.</typeparam>
-        /// <param name="items">The items to prepopulate.</param>
-        /// <returns>The new immutable collection.</returns>
+        /// <typeparam name="T">The type of items to store in the queue.</typeparam>
+        /// <param name="items">The enumerable to copy items from.</param>
+        /// <returns>The new immutable queue.</returns>
         [Pure]
         public static ImmutableQueue<T> CreateRange<T>(IEnumerable<T> items)
         {
             Requires.NotNull(items, nameof(items));
 
-            var queue = ImmutableQueue<T>.Empty;
-            foreach (var item in items)
+            var ilist = items as IList<T>;
+            // Even though ImmutableList supports random access, its indexer has O(log N) time complexity.
+            // So don't employ any IList-based optimizations if `items` is an ImmutableList.
+            if (ilist != null && !(items is ImmutableList<T>))
```

Comment:
> Do we know of cases where CreateRange is used in performance-sensitive code paths with arbitrary ILists?

I would say that most inputs here will probably be ILists, but maybe you're correct that this is more of a convenience method rather than a function someone would use in a perf-sensitive code path. I'll remove the special-casing for IList since the indexer is a black box, and only specialize for arrays (there is a `Create(T[])` method, so we get that almost for free).